### PR TITLE
Option to filter classified data

### PR DESCRIPTION
**Pull request by sduenas**
This PR addresses the requirements defined in #490.

Classified fields can be declared using `CLASSIFIED_FIELDS` class attribute. This attribute will contain a list of lists. Each list stores the "path" or nested dicts keys to the field to filter. For example:
`['my', 'classified', 'field']` will remove `field` from `item['data']['my']['classified']` dict.

To filter this fields from the resulting items, `fetch()` method adds a new parameter `filter_classified`.
Classified fields can also be removed from the command line with the option `filter-classified`.

To start with new feature, this PR adds the field `group.topics` to the list of classified fields in Meetup backend. This field is considered sensible for some related groups working with GrimoireLab.

### CODE REVIEW

File: perceval/backend.py [hunk 1]
```diff
@@ -100,32 +110,48 @@ def archive(self, obj):
...
         :returns: a generator of items
+
+        :raises BackendError: either when the category is not valid or
+            'filter_classified' and 'archive' are active at the same time.
         """
         if category not in self.categories:
             cause = "%s category not valid for %s" % (category, self.__class__.__name__)
             raise BackendError(cause=cause)
 
+        if filter_classified and self.archive:
```
Comment (valeriocos) [inline perceval/backend.py]: The archive stores the url, headers, payload and response of a HTTP request, and the filter is applied later. Why `filter_classified` is disabled when using the archive?

File: perceval/backend.py [hunk 2]
```diff
@@ -161,6 +214,7 @@ def metadata(self, item):
             'origin': self.origin,
             'uuid': uuid(self.origin, self.metadata_id(item)),
             'updated_on': self.metadata_updated_on(item),
+            'classified_fields_filtered': self.classified_fields if filter_classified else None,
```
Comment (valeriocos) [inline perceval/backend.py]: I'm not sure if this should be included to the item here or later only if `filter_classified` is true. Logically, it makes sense to add it where it is now, but in practice it implies that we should re-create all raw indexes we have (to avoid mixing old items created before this change and new ones with the `classified_fields_filtered` attribute)

If we add the field only if `filter_classified` is true, this would only affect the raw indexes for the customer(s) that asked for this feature.

File: perceval/backend.py [hunk 3]
```diff
@@ -192,6 +246,18 @@ def _init_client(self, from_archive=False):
         raise NotImplementedError
 
 
+def _remove_key_from_nested_dict(nested_dict, path_to_field):
```
Comment (valeriocos) [inline perceval/backend.py]: Although it could be a corner case, items with a high nesting level could cause problems in terms of memory consumption. If possible, it would be better to use a loop.

File: tests/test_backend.py [hunk 4]
```diff
@@ -310,6 +360,146 @@ def test_fetch_items(self):
...
+                'item': x,
+                'my': {
+                    'classified': {'field': x},
+                    'field': x,
+                },
+            }
+            self.assertDictEqual(item['data'], expected)
+
+    def test_error_archive_and_filter_classified(self):
+        """CHeck if an error is raised when archive and classified fields filtering are both active"""
```
Comment (valeriocos) [inline tests/test_backend.py]: typo -> CHeck

File: perceval/backend.py (hunk 1 above)
Comment (sduenas) [reply perceval/backend.py]: Because the fields that are classified come on that response. Therefore, they will be stored in the archive which I think it doesn't make sense if you want to remove them from the result. It would cause security problems in your organization, if you don't remember there's an archive storing them.

Comment (valeriocos) [reply perceval/backend.py]: Perfect, could you add a coment on top of that condition ?

Comment (sduenas) [reply perceval/backend.py]: I rather add it to the docstring of that method.

File: perceval/backend.py (hunk 2 above)
Comment (sduenas) [reply perceval/backend.py]: I thought of that but after a long discussion with myself, I still think all items generated with a given version should have the same fields, at least in terms of those fields we control. I know it is verbose for the cases where is not in use, but it allows to see nothing was filtered, instead of wondering why this field is available on these items and not in others generated with the same Perceval version.

In terms of storing these items I would also like to have all items with the same data but sometimes it isn't possible.
On the other hand, GrimoireLab uses a schemaless database (Elastic Search) to store items. This change won't affect already created indexes. New items will be added to existing indexes without problems.

Moreover, take into account, this field only works as reference/debugging. To know if data was filtered from the original or not. It shouldn't be used in queries, etc.

Comment (valeriocos) [reply perceval/backend.py]: Thank you for the clarification, it sounds reasonable.

> On the other hand, GrimoireLab uses a schemaless database (Elastic Search) to store items. This change won't affect already created indexes. New items will be added to existing indexes without problems.

Yes, this change won't affect existing indexes (for the schemaless structure), but we will have items with different meta information. I guess we can live with this discrepancy for a while.

**Comment by sduenas** [comment]
> Overall it looks OK, I left a couple of minor comments. Could you clarify why the filtered items are hard-coded in the backends and cannot be passed as param in the command line ?
> 
> PS fix the flake8 error

Same idea as always. Don't have first and second class items. I'm not saying we shouldn't do that in the future but we don't know yet the implications this feature could cause in the rest of the chain.
What we know for sure is if we allow to do that now, ELK will fail if you filter some of the required fields.

For this iteration, I'd rather stick to this but I'm not saying I didn't think of it. One of the reasons `classified_fields_filtered` in metadata is a list and not a boolean value is this. If in the future we allow to pass a list of fields to filter, the same field will work, storing the fields filtered on those items.

**Comment by valeriocos** [comment]
ok, I was thinking about the case where two customers want to filter different attributes on the same data source. Probably for the first iterations is better to keep it as it is.

File: tests/test_backend.py (hunk 4 above)
Comment (sduenas) [reply tests/test_backend.py]: I think this is one of the cases where recursion works better. Nested fields shouldn't be longer than 4 or 5 keys? I tested with 100 and 1000 nested keys and it works without problems. If wants to filter something with a deep of 10 keys...it's probably because he's doing something terrible wrong.

File: perceval/backend.py (hunk 3 above)
Comment (sduenas) [reply perceval/backend.py]: I think this is one of the cases where recursion works better. Nested fields shouldn't be longer than 4 or 5 keys? I tested with 100 and 1000 nested keys and it works without problems. If wants to filter something with a deep of 10 keys...it's probably because he's doing something terrible wrong.
